src/NodeServices/Interfaces.cs

```diff
@@ -188,4 +191,108 @@ internal interface IExtensionApplication
         /// <param name="baseDirectory"></param>
         void PreloadAsmLibraries(string baseDirectory);
     }
+
+    //TODO should this live in DynamoServices?
+    //make these internal somehow?
+    public interface IExternalCodeCompletionProviderCore
+    {
+
+        /// <summary>
+        /// Already discovered variable types
+        /// </summary>
+        Dictionary<string, Type> VariableTypes { get; set; }
+
+        /// <summary>
+        /// Types that have already been imported into the scope
+        /// </summary>
+        Dictionary<string, Type> ImportedTypes { get; set; }
+
+        /// <summary>
+        /// Generate completion data for the specified text, while import the given types into the
+        /// scope and discovering variable assignments.
+        /// </summary>
+        /// <param name="code">The code to parse</param>
+        /// <param name="expand">Determines if the entire namespace should be used</param>
+        /// <returns>Return a list of IronPythonCompletionData </returns>
+        IExternalCodeCompletionData[] GetCompletionData(string code, bool expand = false);
+
+        /// <summary>
+        /// Try to generate a description from a typename
+        /// </summary>
+        /// <param name="stub">Everything before the last namespace or type name e.g. System.Collections in System.Collections.ArrayList</param>
+        /// <param name="item">Everything after the stub</param>
+        /// <param name="isInstance">Whether it's an instance or not</param>
+        string GetDescription(string stub, string item, bool isInstance);
+
+        bool MatchingEngine(string engineName);
+
+        //TODO think about this, more general?
+        void ImportStdLibrary(string dynamoCorePath);
+
+        void Log(string message);
```

Comment:
We might want to clean this up.
- `Log` does not feel like it should belong here. If needed, it should probably be in another interface.
- `ImportStdLibrary` is not generic enough for a name. Would `Initialize` be a good name?
- How is `MatchingEngine` used? Shouldn't we ask the engine its name, rather than passing one and receiving a `bool`?
- It's not clear to me what `GetDescription` does by looking at its signature.

Reply:
`MatchingEngine` is used to check if this provider can work for a given engine. It needs to be able to be called using reflection. Any thoughts for a different signature?

```
if(CompletionProvider.MatchingEngine(engine.name){
//cool lets use this provider
return CompletionProvider
}
```

Reply:
I think this is resolved.

* Log is removed from the interface and implemented using an existing DynamoCore interface
* ImportStdLibrary is renamed Initialize 
* MatchingEngine I have not changed - open to suggestions.
* GetDescription is a core method needed for the implementation of completion - I'm not yet ready to dive deep and change that - so I'd like to leave it as is, but I did add the summary that I think makes it more clear what it does.